### DISCUSSION TITLE
No Echo in nats jetstream

### DISCUSSION THREAD

**Discussion by anil1996** [nats-io/nats-server#4287] [Q&A] [answered]
Hello,

I have this test code in golang for nats jetstream.

	nats "github.com/nats-io/nats-server/v2/server"
	natsserver "github.com/nats-io/nats-server/v2/test"
	natsio "github.com/nats-io/nats.go"

func TestJSNoEcho(t *testing.T) {
	opts := natsserver.DefaultTestOptions
	opts.NoLog = false
	opts.Port = -1
	opts.JetStream = true
	natsServer = natsserver.RunServer(&opts)
	msgAddr := natsServer.Addr().String()
	nc, err := natsio.Connect(msgAddr, natsio.NoEcho())
	require.Nil(t, err)
	js, err := nc.JetStream()
	require.Nil(t, err)
	require.NotNil(t, js)
	_, err = js.AddStream(&natsio.StreamConfig{
		// name of the stream
		Name:     "ORDERS11",
		Subjects: []string{"ORDERS11"},
		Storage:  natsio.MemoryStorage,
	})
	require.Nil(t, err)
	sub, err := js.Subscribe("ORDERS11", func(msg *natsio.Msg) {
		t.Logf("message received in subscribe %v", string(msg.Data))
	})
	require.NotNil(t, sub)
	require.Nil(t, err)
	pub, err := js.Publish("ORDERS11", []byte("hello"))
	require.NotNil(t, pub)
	require.Nil(t, err)
	time.Sleep(5 * time.Second)

}

I am expecting that since I am using same client for publish and subscribe and used no echo option. I should not be getting this message on subscription. But I am still getting that message meaning the log that I have inside handler for subscription prints.

It will be really helpful if someone can help me identify where its going wrong. I want to stop echoing using jetstream.
Thanks

**Comment by aricart** [comment, marked as answer]
Your issue is that you are publishing on nats core (where the no-echo) implements functionality - but you are creating a jetstream consumer that is processing messages from a stream that captures the message you are publishing. In JetStream you are processing messages from a stream, the fact that you also published that message is not known to JetStream and no-echo has no effect on that.

**Comment by anil1996** [comment]
Hi @aricart ,
Thanks for the quick response.
I am using jetstream context for publish. Is it not different than nats core publish?
Also if I have to fix this issue, how can I instruct jetstream to take no-echo option in consideration? Any example would be appreciated.
Thanks

**Comment by aricart** [comment]
the message you are publishing is through nats core, the JS publish functionality is simply making it a request so you can receive confirmation that your message was received by the server and written to the stream.

Your subscription, however (even though the function is called `Subscribe()` is really creating a consumer, finding the stream by looking for streams that capture the subject you published, creating a nats subscription (in your case a push subscription), and then the server is forwarding messages from that stream, to the push subscription which you are getting) - what you are doing is really playing back the log of messages, JetStream doesn't have any idea you published that message

**Reply by aricart** [reply]
The important detail that may be overlooked is that the nats subscription that is getting the messages is NOT on the subject that you published, but a different generated subject - jetstream is simply sending protocol and data messages to that subscription which then the client processes - the subject you are specifying in the subscribe is really a "filter" on the stream, and is used for identifying the stream that you want to get messages from, and filtering on subjects that match your subscribe.

**Comment by aricart** [comment]
Also, since you seem to be getting started with JetStream, you should really look at the consumer functionality in the latest releases of the nats.go client, that will make your code simpler. See https://github.com/nats-io/nats.go/blob/main/jetstream/README.md
